packages/go/analysis/ad/esc13.go

```diff
@@ -112,3 +152,263 @@ func getCertTemplateGroupLinks(ct *graph.Node, tx graph.Transaction) (graph.Node
+		Root: startNode,
+		Driver: adcsESC13Path2Pattern(cardinality.DuplexToGraphIDs(path1EnterpriseCAs), cardinality.DuplexToGraphIDs(path1DomainNodes)).Do(func(terminal *graph.PathSegment) error {
+			var enterpriseCANode *graph.Node
+			terminal.WalkReverse(func(nextSegment *graph.PathSegment) bool {
+				if nextSegment.Node.Kinds.ContainsOneOf(ad.EnterpriseCA) {
+					enterpriseCANode = nextSegment.Node
+				}
+				return true
+			})
+			lock.Lock()
+			path2CandidateSegments[enterpriseCANode.ID] = append(path2CandidateSegments[enterpriseCANode.ID], terminal)
+			path2EnterpriseCAs.Add(enterpriseCANode.ID.Uint32())
+			lock.Unlock()
+
+			return nil
+		}),
+	}); err != nil {
+		return nil, err
+	}
+
+	//Manifest P3 keyed to cert template nodes
+	if err := traversalInst.BreadthFirst(ctx, traversal.Plan{
+		Root: endNode,
+		Driver: adcsESC13Path3Pattern(cardinality.DuplexToGraphIDs(path1CertTemplates)).Do(func(terminal *graph.PathSegment) error {
+			certTemplate := terminal.Search(func(nextSegment *graph.PathSegment) bool {
+				return nextSegment.Node.Kinds.ContainsOneOf(ad.CertTemplate)
+			})
+
+			lock.Lock()
+			path3CandidateSegments[certTemplate.ID] = append(path3CandidateSegments[certTemplate.ID], terminal)
+			lock.Unlock()
+			return nil
+		}),
+	}); err != nil {
+		return nil, err
+	}
+
+	//Manifest P4, keyed to the domain nodes
+	if err := traversalInst.BreadthFirst(ctx, traversal.Plan{
+		Root: endNode,
+		Driver: adcsESC13Path4Pattern().Do(func(terminal *graph.PathSegment) error {
+			if terminal.Node.Kinds.ContainsOneOf(ad.Domain) {
+				lock.Lock()
+				path4CandidateSegments[terminal.Node.ID] = append(path4CandidateSegments[terminal.Node.ID], terminal)
+				lock.Unlock()
+			}
+
+			return nil
+
+		}),
+	}); err != nil {
+		return nil, err
+	}
+
+	//And the 2 bitmaps together to ensure that only enterprise CAs thats satisfy both paths are valid
+	path1EnterpriseCAs.And(path2EnterpriseCAs)
+
+	for k, v := range path1CandidateSegments {
+		path2segments := path2CandidateSegments[k]
+		for _, p1 := range v {
+			for _, p2 := range path2segments {
+				//Check if our terminal nodes match (should be the same domain node)
+				if p1.Node.ID != p2.Node.ID {
+					continue
+				}
+
+				//Find the cert template in path 1 and use that to find the correct p3 segment
+				certTemplate := p1.Search(func(nextSegment *graph.PathSegment) bool {
+					return nextSegment.Node.Kinds.ContainsOneOf(ad.CertTemplate)
+				})
+
+				if p3segments, ok := path3CandidateSegments[certTemplate.ID]; !ok {
+					continue
+				} else if p4segments, ok := path4CandidateSegments[p1.Node.ID]; !ok {
+					//If we dont have a domain contain relationship, then this is not a valid path
+					continue
+				} else {
+					//Merge all our paths together
+					paths.AddPath(p1.Path())
+					paths.AddPath(p2.Path())
+					for _, p3 := range p3segments {
+						paths.AddPath(p3.Path())
+					}
+					for _, p4 := range p4segments {
+						paths.AddPath(p4.Path())
+					}
```

Comment:
Sorry for extra comment after review:
I just realized that this for loop is also not needed once it's done at least one time right? Not sure if we can pop this outside of the path2segment loop as well or if there's a way we can skip it if it's already been done once.

Also not sure if all this optimization is necessary either so if it's too hyper optimized, feel free to ignore as well.

Reply:
I'm going to leave it as is to make it abundantly clear what the conditions are for creating paths  that fully manifest. Otherwise its possible for the paths slice to be populated with incomplete paths